### PR TITLE
VIRTS-838/Implement DNS C2

### PR DESCRIPTION
**Pull request by christophert**
-- WIP --

# DNS C2

# Data Format
ALL DATA WILL BE ZLIB COMPRESSED ON BOTH SIDES

## Encode
```python
str(base64.urlsafe_b64encode(ZLIB.compress(s.encode(), 9)).decode())
```

## Decode
```python
str(zlib.decompress(base64.urlsafe_b64decode(s)).decode().replace('\n', ''))
```

# Ping

```
REQ ping.54ndc47.<suffix>. A
RESP A 80.79.78.71
```

```
FORMAT OF REQ = <base64 / 100 OR length of transmission>.<sequence number 0x0000/0xFFFF *ONLY* in requests>.<request type (01/02)>.<tid if in transmission>.<[s]tart/[d]one/[c]ontinue/[r]eceive/[r]eceive[d]one>.<paw>.54ndc47.<suffix>
```

# Instructions = 01
## Initiate Transmission
```
REQ 41414140.01.s.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"success": "true", "tid": <tid>}>
```

## Transfer Data
```
REQ <base64 / 100>.<seq number 0x0000/0xFFFF>.01.<tid>.c.<paw>.54ndc47.<suffix> A/CNAME
RESP A 41.41.41.41 = ACK

...

REQ <length of transmission>.01.<tid>.d.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"success": "true", "data": <data>}>
OR (if chunking is not required)
RESP TXT <base64 {"success": "false", "error": "error"}>
OR (if response needs to be chunked due to DNS size limit)
RESP TXT <base64 {"success": "true", "chunked": true, "total_chunks": # total chunk number}>
```

If response data is chunked, client will request data from the server as follows:

```
REQ <seq number>.01.<tid>.r.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"chunk": #chunk, "data": <data>}>
....
WHEN DONE WITH CHUNK TRANSFER
REQ 01.<tid>.rd.<paw>.54ndc47.<suffix> A
RESP A 41.41.41.41 ACK
```

# Results = 02
```
REQ 41414140.02.s.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"success": "true", "tid": <tid>}>

REQ <base64 of requests / 100>.<seq number>.02.<tid>.c.<paw>.54ndc47.<suffix> A
RESP A 41.41.41.41 = ACK

REQ <length of transmission>.02.<tid>.d.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"success": "true", "data": <data>}>
OR
RESP TXT <base64 {"success": "false", "error": "error"}>
```

# Download File = 03
## Initiate Transmission
```
REQ 41414140.03.s.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"success": "true", "tid": <tid>}>
```

## Transfer Data
```
request = encode({"file": "sandcat.go", "platform": "linux", ... })
REQ <base64 of request / 100>.<seq number>.03.<tid>.c.<paw>.54ndc47.<suffix> A/CNAME
RESP A 41.41.41.41 = ACK

...

REQ <length of transmission>.03.<tid>.d.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"success": "true", "data": <data>}>
OR (if chunking is not required)
RESP TXT <base64 {"success": "false", "error": "error"}>
OR (if response needs to be chunked due to DNS size limit)
RESP TXT <base64 {"success": "true", "chunked": true, "total_chunks": # total chunk number}>
```

If response data is chunked, client will request data from the server as follows:

```
REQ <seq number>.03.<tid>.r.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"chunk": #chunk, "data": <data>}>
....
WHEN DONE WITH CHUNK TRANSFER
REQ 03.<tid>.rd.<paw>.54ndc47.<suffix> A
RESP A 41.41.41.41 ACK
```

# Upload File = 04
```
REQ 41414140.04.s.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"success": "true", "tid": <tid>}>

request = encode({"filename": "drop.off", "contents": encode(binary data)})
REQ <base64 of requests / 100>.<seq number>.04.c.<paw>.53ndc47.<suffix> A
RESP A 41.41.41.41 ACK

REQ <length of transmission>.<tid>.04.d.<paw>.54ndc47.<suffix> TXT
RESP TXT <base64 {"success": "true"}>
OR
RESP TXT <base64 {"success": "false", "error": "error"}>
```

### CODE REVIEW

File: app/contact/dns.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+        elif req_type == 3:
+            success, ret = await self._download_file(data)
+        elif req_type == 4:
+            success, ret = await self._upload_file(paw, data)
+        else:
+            success, ret = False, "Invalid request type"
+        return success, ret
+
+    async def resolve(self, request, handler):
+        """
```
Comment (khyberspache) [inline app/contact/dns.py]: This function is way to large - let's start by breaking each of the ELIFs into individual functions.

File: app/contact/dns.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+        resp = request.reply()
+        if isinstance(rrs, list):
+            for rr in rrs:
+                resp.add_answer(rr)
+        else:
+            resp.add_answer(rrs)
+        return resp
+
+    @staticmethod
+    def decode_bytes(s):
```
Comment (khyberspache) [inline app/contact/dns.py]: Are these meant to be private?

File: app/contact/dns.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        """
+        ret = []
+        while len(data) > 0:
+            ret.append(data[:chunk_size])
+            del data[:chunk_size]
+        return ret
+
+
+class DNS(C2Passive):
+    """
```
Comment (khyberspache) [inline app/contact/dns.py]: Move this out of the Private function section.  I think this should be first class in the file...

**Comment by christophert** [comment]
Moved internal

**Pull request closed without merging**